python/mlc_chat/compiler/model/phi/phi_model.py

```diff
@@ -0,0 +1,329 @@
+        self.rotary_embedding = rotary_embedding
+        self.n_head = config.n_head
+        self.n_head_kv = config.n_head_kv
+        self.head_dim = config.head_dim
+        op_size = self.head_dim * (self.n_head + 2 * self.n_head_kv)
+        hidden_size = config.n_embd
+
+        self.Wqkv = nn.Linear(hidden_size, op_size, bias=True)
+        self.out_proj = nn.Linear(hidden_size, hidden_size, bias=True)
+        self.inner_cross_attn = PhiCrossAttention(config)
+        self.k_cache = nn.KVCache(config.context_window_size, [self.n_head_kv, self.head_dim])
+        self.v_cache = nn.KVCache(config.context_window_size, [self.n_head_kv, self.head_dim])
+
+    def forward(self, x: Tensor, attention_mask: Tensor, total_seq_len: tir.Var):
+        d, h_q, h_kv, t = self.head_dim, self.n_head, self.n_head_kv, total_seq_len
+        b, s, _ = x.shape
+        assert b == 1, "Only support batch size 1 at this moment."
+        # Step 1. QKV Projection
+        qkv = self.Wqkv(x)
+        qkv = op.reshape(qkv, (b, s, h_q + h_kv + h_kv, d))
+        q, k, v = op.split(qkv, indices_or_sections=[h_q, h_q + h_kv], axis=2)
+        # Step 2. Apply QK rotary embedding
+        q, k = self.rotary_embedding(q, k, t - s)
+        # Step 3. Query and update KVCache
+        self.k_cache.append(op.squeeze(k, axis=0))
+        self.v_cache.append(op.squeeze(v, axis=0))
+        k = self.k_cache.view(t)
+        v = self.v_cache.view(t)
+        # Step 4. Compute softmax(Q @ K^T / sqrt(d)) @ V
+        output = self.inner_cross_attn(q, k, v, attention_mask)
+        # Step 5. Apply output projection
+        return self.out_proj(output)
+
+
+class PhiParallelBlock(nn.Module):
+    def __init__(self, config: PhiConfig, rotary_embedding: RotaryEmbedding):
+        super().__init__()
+
+        self.ln = nn.LayerNorm(config.n_embd, eps=config.layer_norm_epsilon)
+        self.mixer = PhiMHA(config, rotary_embedding)
+        self.mlp = PhiMLP(config)
+
+    def forward(self, hidden_states: Tensor, attention_mask: Tensor, total_seq_len: tir.Var):
+        residual = hidden_states
+        hidden_states = self.ln(hidden_states)
+
+        attn_outputs = self.mixer(
+            hidden_states,
+            attention_mask,
+            total_seq_len,
+        )
+
+        feed_forward_hidden_states = self.mlp(hidden_states)
+
+        hidden_states = attn_outputs + feed_forward_hidden_states + residual
+
+        return hidden_states
+
+
+class PhiCausalLMHead(nn.Module):
+    def __init__(self, config: PhiConfig) -> None:
+        super().__init__()
+
+        self.ln = nn.LayerNorm(config.n_embd, eps=config.layer_norm_epsilon)
+        self.linear = nn.Linear(config.n_embd, config.vocab_size)
+
+    def forward(self, hidden_states: Tensor):
+        hidden_states = self.ln(hidden_states)
+        logits = self.linear(hidden_states)
+
+        if logits.dtype != "float32":
+            logits = logits.astype("float32")
```

Comment:
do we always cast logits to fp32?

Reply:
yes, we add the casting to most of the models